### PR TITLE
feat: make typst-preview as a library

### PR DESCRIPTION
**Pull request by Myriad-Dreamin**
The interface becomes:

```rs
pub async fn preview<T: CompileHost + Send + 'static>(
    arguments: PreviewArgs,
    client: impl FnOnce(CompilationHandleImpl) -> T,
) -> Previewer;
```

The change since last PR is the second argument of this function:

First, it provides a compilation handle for creating a compilation host:

```rs
impl CompilationHandle for CompilationHandleImpl {
    fn status(&self, status: CompileStatus) {
        self.editor_tx...
    }

    fn notify_compile(&self, res: Result<Arc<Document>, CompileStatus>) {
        match res {
            Ok(doc) => {
                ...
            }
            Err(status) => {
                ...
            }
        }
    }
}
```

Second, caller creates two callback, all of whose methods are optional:

```rs
pub trait SourceFileServer {
    fn resolve_source_span(
        &mut self,
        _by: Location,
    ) -> impl std::future::Future<Output = ZResult<Option<SourceSpanOffset>>> + Send {
        async { Ok(None) }
    }

    fn resolve_document_position(
        &mut self,
        _by: Location,
    ) -> impl std::future::Future<Output = ZResult<Option<Position>>> + Send {
        async { Ok(None) }
    }

    fn resolve_source_location(
        &mut self,
        _s: Span,
        _offset: Option<usize>,
    ) -> impl std::future::Future<Output = ZResult<Option<DocToSrcJumpInfo>>> + Send {
        async { Ok(None) }
    }
}

pub trait EditorServer {
    fn update_memory_files(
        &mut self,
        _files: MemoryFiles,
        _reset_shadow: bool,
    ) -> impl std::future::Future<Output = ZResult<()>> + Send {
        async { Ok(()) }
    }

    fn remove_shadow_files(
        &mut self,
        _files: MemoryFilesShort,
    ) -> impl std::future::Future<Output = ZResult<()>> + Send {
        async { Ok(()) }
    }
}
```

The `src/main.rs` and `src/compiler.rs` could provides a sample implementation.

### CODE REVIEW

File: src/actor/typst.rs [hunk 1]
```diff
@@ -264,26 +174,35 @@ impl TypstClient {
                     "TypstActor: processing SYNC memory files: {:?}",
                     m.files.keys().collect::<Vec<_>>()
                 );
-                self.update_memory_files(m, true);
+                handle_error(
+                    "SyncMemoryFiles",
+                    self.client.update_memory_files(m, true).await,
```
Comment (Enter-tainer) [inline src/actor/typst.rs:179]: plz instrument all awaits

File: src/compiler.rs [hunk 2]
```diff
@@ -0,0 +1,187 @@
...
+
+        Self {
+            inner,
+            client: TypstClient {
+                inner: once_cell::sync::OnceCell::new(),
+            },
+        }
+    }
+
+    pub fn spawn(self) -> ZResult<TypstClient<H>> {
```
Comment (Enter-tainer) [inline src/compiler.rs:94]: what is ZResult?

File: src/compiler.rs [hunk 3]
```diff
@@ -0,0 +1,187 @@
...
+impl<H: CompilationHandle> TypstClient<H> {
+    fn inner(&mut self) -> &mut CompileClient<H> {
+        self.inner.get_mut().unwrap()
+    }
+}
+
+impl<H: CompilationHandle> SourceFileServer for TypstClient<H> {
+    async fn resolve_source_span(&mut self, loc: Location) -> ZResult<Option<SourceSpanOffset>> {
+        let Location::Src(src_loc) = loc;
+        self.inner().resolve_src_location(src_loc).await
```
Comment (Enter-tainer) [inline src/compiler.rs:117]: same, plz instrument

File: src/lib.rs [hunk 4]
```diff
@@ -88,65 +132,100 @@ impl Previewer {
...
+
+pub enum Location {
+    Src(SourceLocation),
+}
+
+pub trait SourceFileServer {
+    fn resolve_source_span(
+        &mut self,
+        _by: Location,
+    ) -> impl std::future::Future<Output = ZResult<Option<SourceSpanOffset>>> + Send {
```
Comment (Enter-tainer) [inline src/lib.rs:145]: use std::future::Future is better

**Review by Enter-tainer — APPROVED** [review]
generally lgtm

File: src/compiler.rs (hunk 2 above)
Comment (Myriad-Dreamin) [reply src/compiler.rs:94]: `ZResult` is defined in typst.ts/core/error. It is `Result<T, typst_ts_core::Error>`

Comment (Enter-tainer) [reply src/compiler.rs:94]: hmm.. The name is not good. anyway, it's not a big deal

Comment (Myriad-Dreamin) [reply src/compiler.rs:94]: I don't like `type Result<T> = std::Result<T, ts::Error>`, neither `typst TsResult<T> = ...`, so I add one extra alphabet to reduce redundancy.

Comment (Myriad-Dreamin) [reply src/compiler.rs]: I have inlined `ZResult` in typst_preview. it is heavily used in ts but not preview, so we can tolerate the redundancy.

**Pull request merged**